DPI related bug-fixes

shubham-cdot opened the pull request.

shubham-cdot committed:
DPI: remove zero data length check from dpi_ndpi_session_first_packet()

For TCP sessions, L4 payload data length of first packet
(SYN packet) is zero. So the first packet of TCP session
will not be sent to nDPI due to data_len != 0 checking in
dpi_ndpi_session_first_packet().

However, nDPI uses TCP SYN packets internally for connection
tracking and other purposes. For example, in nDPI library
function ndpi_detection_process_packet():
a) ndpi_connection_tracking() is called for connection
tracking and updating TCP flag states
b) app detection is given up for TCP sessions in some cases if
the first packet of session sent to nDPI is not a SYN packet

Hence, remove data_len != 0 check from the function
dpi_ndpi_session_first_packet() and send the first packet of
a session to nDPI irrespective of data length.

Co-authored-by: Subhajit Chatterjee <subhajit.chatterjee@cdot.in>
Signed-off-by: Shubham Shrivastava <shubhams@cdot.in>

shubham-cdot committed:
DPI: use USE_NDPI flag

Add USE_NDPI flag to initialise ‘engines’ and 'engines_len'
before calling dpi_session_first_packet() function in order
to avoid error due to missing dpi_engine_procs for nDPI engine
when dataplane compilation is done without USE_NDPI flag.

Co-authored-by: Subhajit Chatterjee <subhajit.chatterjee@cdot.in>
Signed-off-by: Shubham Shrivastava <shubhams@cdot.in>

shubham-cdot committed:
DPI: use correct loop variable

Use loop variable 'j' instead of 'i' during flow cleanup.

Co-authored-by: Subhajit Chatterjee <subhajit.chatterjee@cdot.in>
Signed-off-by: Shubham Shrivastava <shubhams@cdot.in>

pjaitken commented:
@shubham-cdot could you tell us why you had to remove the zero data length check from dpi_ndpi_session_first_packet?

Are there some applications which are not being identified? Is this particular to 3.4?

Can you give us details of, "app detection is given up for TCP sessions in some cases if the first packet of session sent to nDPI is not a SYN packet"?

I'm trying to understand why this wasn't an issue for us before, and I'd appreciate any information which we could use to improve our internal DPI testing.

Thanks.

shubham-cdot commented:
@pjaitken In ndpi_detection_process_packet function (file: ndpi_main.c):

>  if((flow->num_processed_pkts == 1) && 
    (ret.master_protocol == NDPI_PROTOCOL_UNKNOWN) &&
     (ret.app_protocol == NDPI_PROTOCOL_UNKNOWN) && 
    flow->packet.tcp && 
     (flow->packet.tcp->syn == 0) &&
    (flow->guessed_protocol_id == 0)) {
	    u_int8_t protocol_was_guessed;

	    /*
	      This is a TCP flow
	      - whose first packet is NOT a SYN
	      - no protocol has been detected

	      We don't see how future packets can match anything
	      hence we giveup here
	    */
	    ret = ndpi_detection_giveup(ndpi_str, flow, 0, &protocol_was_guessed);
>  }
 

The above checking was first added to nDPI via commit id e4f01976a and is present since nDPI version 3.0:

>commit e4f01976a66f1943bde7b253b62430d36c6d9e74
Author: Luca <deri@ntop.org>
Date:   Thu Aug 30 11:10:30 2018 +0200

    Added missing categorization when giveup/guess is called
    Added optimization for TCP flows that do not start with a SYN packet: early giveup is performed
    Code cleanup


As per our understanding, app detection will be given up for a TCP flow as per above checking in case of TCP flows for which nDPI is unable to detect the app protocol using the first packet sent to nDPI after the SYN packet, by applying both: 
* protocol guessing (for example based on IP address, TCP ports etc.) and
* protocol dissectors

This would be more likely to happen for protocols running on non-standard ports for which multiple packets of the flow are required for app identification. 

We tried to simulate this scenario but currently unable to do so due to limitations of test setup at our end. For more info, please refer to packet handling in ndpi_detection_process_packet. 
Thanks!

pjaitken commented:
@shubham-cdot, I'm copying some internal feedback from @dfawcus.

pjaitken commented on src/npf/dpi/ndpi.c:
```diff
@@ -315,8 +316,7 @@ dpi_ndpi_session_first_packet(struct npf_session *se __unused,
 		goto dest_id_error;
 
-	if (data_len != 0 && !dpi_ndpi_process_pkt(
-				(struct dpi_engine_flow *)flow, mbuf, dir))
```
@dfawcus

In dpi.c:dpi_get_data_len() there is a comment around where the UDP length is validated, intended to simply cause packets to be dropped if the length is invalid.

The above change would seem to make the call graph to here broken, in that such packets now will not be dropped, so there would need to be another way of dealing with that case.

It is also worth keeping in mind that (eventually) UDP packets will have the possibility of having 0 traditional data, while actually carrying elsewhere because of the UDP-Options draft which is in progress.

It strikes me that the data_length field should maybe be removed from the prototype for this function, as it arose when the mbuf was not passed in, and we passed instead a data+length tuple. It is rather superfluous at the moment, as it can be extracted from the mbuf - see dpi_get_data_length().

pjaitken commented on src/npf/rproc/npf_ext_appfw.c:
```diff
@@ -389,10 +389,15 @@ appfw_action(npf_cache_t *npc, struct rte_mbuf **nbuf, void *arg,
+#else
+		uint8_t engines[] = { IANA_USER };
+		size_t engines_len = 1;
+#endif /* USER_NDPI */
```
@dfawcus

Not directly related, but possibly important.

I believe that when the app-fw was created, 0 data-length packets were not fed through it, and so did not count towards the number of initial packets it had to see before converging on an answer. So TCP flags only updates would not count.

The context of this PR implies that this is no longer the case, and it may inform upon why we had issues with the initial count behaviour in the post Qosmos world. So there may be something to be updated in the app-fw now as to which packets it counts towards that initial set - possibly 0-data length TCP packet should not be counted?

dfawcus replied:
Possibly copied to the wrong place, this makes more sense in relation to the prior commit.
(I can't recall which commit I added the comment to in the internal PR system).

pjaitken commented on src/npf/rproc/npf_ext_appfw.c:
```diff
@@ -389,10 +389,15 @@ appfw_action(npf_cache_t *npc, struct rte_mbuf **nbuf, void *arg,
+		size_t engines_len = 2;
+#else
+		uint8_t engines[] = { IANA_USER };
+		size_t engines_len = 1;
```
@dfawcus

The second line would be better outwith of the conditional and written as:

    size_t engines_len = ARRAY_SIZE(engines);

Which raises another question, each other call to dpi_session_first_packet() passes an final pair or parameters, being 'engines_length' and 'engines', however 'engines_length' was a constant '2' in each case.

This change implies that something related to that pair should be altered - somehow.

pjaitken commented on src/pipeline/nodes/l3_dpi.c:
```diff
@@ -75,8 +75,15 @@ ip_dpi_process_common(struct pl_packet *pkt, bool v4, int dir)
+		size_t engines_len = 2;
+#else
+		uint8_t engines[] = {IANA_USER};
+		size_t engines_len = 1;
```
@dfawcus

ARRAY_SIZE, as earlier.

pjaitken commented on src/npf/rproc/npf_ext_dpi.c:
```diff
@@ -178,9 +178,15 @@ dpi_match(npf_cache_t *npc, struct rte_mbuf *mbuf, const struct ifnet *ifp,
+		size_t engines_len = 2;
+#else
+		uint8_t engines[] = {IANA_USER};
+		size_t engines_len = 1;
```
@dfawcus

ARRAY_SIZE, as earlier.

pjaitken replied:
I think this repeated code should be in a little helper function so USE_NDPI is kept to a minimum, and there's only one place to add additional DPI engines in future.

dfawcus commented:
@dfawcus-vyatta

dfawcus-vyatta commented:
@dfawcus

dfawcus commented:
@dfawcus

subhajit-cdot commented:
@pjaitken What if we remove the check for UDP packets from dpi.c:dpi_get_data_len() ? How will it impact the DPI feature? The issue might arise for some protocol detection (TCP based) running on non standard ports. The nDPI might give up if data_len == 0 packets are skipped.

We can use ARRAY_SIZE() to find num of engines rather than static entry , however can you elaborate  "there's only one place to add additional DPI engines in future." This is true if we use num of engines = 2 ?

Regards,
S

dfawcus commented:
> What if we remove the check for UDP packets from dpi.c:dpi_get_data_len() ? How will it impact the DPI feature? The issue might arise for some protocol detection (TCP based) running on non standard ports. The nDPI might give up if data_len == 0 packets are skipped.

My point is that the code I referred to needs a mechanism to force some packets to be dropped.  The code originally made use of indicating the length was zero to achieve that.  If now a length of zero can not be used to signal that the packet should be dropped, then some other mechanism needs to be provided.

I am not suggesting disallowing a length zero packet to processed, I am suggesting that you need to add a alternate mechanism to force some packets to be dropped.

pjaitken commented:
@subhajit-cdot, I mean that engines[] and engines_len should be set in only one place (ie, in a new function) rather than in multiple places throughout the code.
If a new engine DPI engine is added then it will only need to be added in that one function rather than in multiple places.

shubham-cdot commented:
> My point is that the code I referred to needs a mechanism to force some packets to be dropped. The code originally made use of indicating the length was zero to achieve that. If now a length of zero can not be used to signal that the packet should be dropped, then some other mechanism needs to be provided.

In order to distinguish invalid data length case from 0 (valid) data length, we can do one of the following:
1. add an error flag to dpi.c:dpi_get_data_len() as an output argument to indicate invalid data length case
2. add new function just to check for invalid data length

In case of option 2, the checking for invalid data length will be moved from dpi_get_data_len() to the new function.

shubham-cdot commented:
@pjaitken @dfawcus Few queries:

a) In the original DANOS 2009 code, the check for data_len != 0 is done only in   ndpi.c:dpi_ndpi_session_first_packet() i.e. for the first packet of session. Will this check not be required for subsequent packets of the session? 

b) In dpi.c:dpi_get_data_len(), data length is validated only for UDP. Will similar check not be required for TCP?

subhajit-cdot commented:
Hello @pjaitken 
Can you please update about the current status of the two pending bug fixes (data_len and engines)?
Thanks
Subhajit

pjaitken commented:
@shubham-cdot, @subhajit-cdot,

The data_len != 0 check should be removed. All valid packets should be sent to the DPI engine.

Should packets with an invalid length be sent to the DPI engine or not? If the DPI engine validates packets itself, and does not overrun the packet or crash, then we can simply send all traffic into the engine. If we don't send invalid length packets to the DPI engine, then should further packets be sent to the engine after an invalid length packet has been blocked?

If we agree to block invalid length packets, then a new "invalid length" check should be added - for all packets (not just the first) and for all protocols (not just UDP). It seems sensible to add the invalid length check to dpi_get_data_len and return an output argument. It would be useful to return the length of other transport protocols too, though this could be done later.

The DPI statistics should only count packets which are being sent to the DPI engine.

Regarding engines, engines[] and engines_len should be set in a new function so this is done in only one place rather than in multiple places throughout the code.

Thanks.
